src/mlpack/methods/preprocess/preprocess_string_encoding_main.cpp

```diff
@@ -0,0 +1,245 @@
+            else
+              fout << encodedResult.at(j)(i, k);
+          }
+        }
+      }
+      else
+      {
+        if (j + 1 < dataset[i].size())
+          fout << dataset[i][j] << columnDelimiter;
+        else
+          fout << dataset[i][j];
+      }
+    }
+    fout << "\n";
+  }
+}
+
+static void mlpackMain()
+{
+  // Parse command line options.
+  // Extracting the filename
+  const string filename = CLI::GetParam<string>("actual_dataset");
+  string columnDelimiter;
+  if (CLI::HasParam("column_delimiter"))
+  {
+    columnDelimiter = CLI::GetParam<string>("column_delimiter");
+    // Allow only 3 delimiters.
+    RequireParamValue<string>("column_delimiter", [](const string del)
+        { return del == "\t" || del == "," || del == " "; }, true,
+        "Delimiter should be either \\t (tab) or , (comma) or ' ' (space) ");
+  }
+  else
+    columnDelimiter = data::ColumnDelimiterType(filename);
+
+  // Handling Dimension vector
+  vector<string> tempDimension =
+      CLI::GetParam<vector<string> >("dimension");
+  unordered_set<size_t> dimensions = data::GetColumnIndices(tempDimension);
+  vector<vector<string>> dataset = data::CreateDataset(filename,
+      columnDelimiter[0]);
+  for (auto colIndex : dimensions)
+  {
+    if (colIndex >= dataset.back().size())
+      Log::Fatal << "The index given is out of range, please verify" << endl;
+  }
+
+  // Preparing the input dataset on which string manipulation has to be done.
+  // vector<vector<string>> nonNumericInput(dimension.size());
+  unordered_map<size_t , vector<string>> nonNumericInput;
+
+  for (size_t i = 0; i < dataset.size(); i++)
+  {
+    for (auto& datasetCol : dimensions)
+      nonNumericInput[datasetCol].push_back(move(dataset[i][datasetCol]));
+  }
+
+  unordered_map<size_t , arma::mat> encodedResult;
+  const string delimiters = CLI::GetParam<string> ("delimiter");
+  data::SplitByAnyOf tokenizer(delimiters);
+
+
+  if (CLI::HasParam("encoding_type"))
+  {
+    RequireParamValue<string>("encoding_type", [](const string et)
+        { return et == "DictionaryEncoding" || et == "BagOfWordsEncoding"
+        || et == "TfIdfEncoding"; }, true, "Encoding Type should be either"
+        " BagOfWordsEncoding or DictionaryEncoding or TfIdfEncoding ");
+  }
+  const string& encodingType = CLI::GetParam<string>("encoding_type");
+  arma::mat output;
+  using TokenType = data::SplitByAnyOf::TokenType;
+  for (auto& column : nonNumericInput)
+  {
+    if (encodingType == "DictionaryEncoding")
+    {
+        // dictionary Encoding.
+      data::DictionaryEncoding<TokenType> encoder;
+      encoder.Encode(column.second, output, tokenizer);
+      encodedResult[column.first] = std::move(output);
+    }
+    else if (encodingType == "BagOfWordsEncoding")
+    {
+      // BagofWords Encoding.
+      data::BagOfWordsEncoding<TokenType> encoder;
+
+      encoder.Encode(column.second, output, tokenizer);
+      encodedResult[column.first] = std::move(output);
```

Comment:
You could use templates here to simplify the code; basically you could just have a helper function `EncodeColumn` that takes a template parameter `EncodingType`, and then you'd just set `EncodingType` in each block of this `if`, e.g., `data::BagOfWordsEncoding<TokenType>`.

Reply:
I was wondering if it has any sense to implement a model class for this CLI utility. On the other hand, this utility isn't that complicated, and the model class could sufficiently increase the amount of code instead of simplify it.